simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/Registry.java

```diff
@@ -100,7 +100,7 @@ private Registry() {}
   public static Set<Policy> policies(BasicSettings settings, Set<Characteristic> characteristics) {
     return settings.policies().stream()
         .flatMap(name -> policy(settings, name).stream())
-        .filter(policy -> policy.characteristics().containsAll(characteristics))
+        .filter(policy -> characteristics.containsAll(policy.characteristics()))
```

Comment:
Why is this inverted?

The prior logic said that the policy must implement the minimum characteristics, as defined by the trace. This allowed the policy to be weighted and the trace not be, so the default weight of `1` would be supplied and compatible.

The new logic says that the policy must implement a subset of the trace's characteristics.

Reply:
I see what you're saying, but I had my extension of a trace with penalties, where I would like for policies that do not use this characteristic to still run, so it will be possible to see how they compare by the metric of QoS/latency.
a possibility to handle this issue can be to add an attribute to the config file that will decide if it should be: the prior logic, the new logic, and maybe another option of equality of the sets.

BTW, a check if the number of policies that remain after filtering in the registry is > 0 is needed, else the simulator won't stop. (I'll add this now)

let me know what you think about my suggestion regarding the config attribute to handle the filtering logic, so it would be able to fit multiple simulation requirements.

Reply:
It sounds like this isn't an optional characteristic then, because you want all policies to report latency information. This means that, like `key`, it is fundamental to a policy to handle. What is the benefit of having the characteristic then? It merely says whether the algorithm uses it in its decisions, which is not very meaningful (as they don't report whether they use recency or frequency, for instance). If that's the case, then I don't think you should add a characteristic.

This would explain why the changes were more invasive than I expected. It would be nice to not have it impact every one written, though perhaps there is no workaround. The latencies could change over the lifetime workload (e.g. network getting slower when congested), which could cause some surprising skew based on when the miss occurred.

Reply:
One approach might be to add the penalty in `PolicyActor`, like we do for the execution time. If the miss count increased, add the miss penalty. If the hit count increased, that one. If neither, ignore (e.g. opt must wait for its replay). That's a little less invasive as writing a policy won't need to account for the QoS metrics.

Reply:
Right now, there are no implemented policies that use this information for decisions, but I intend on adding such a policy, perhaps even more than one.
And, when using it I might want to compare it's stats to other policies that might not use this information in the decision-making process.

regarding the changes in the miss penalties over time, I am aware of that, and the algorithm I'm working on tries to take that into consideration, in addition, we are now working on traces from a real system. 

so if a data from the trace is needed for some of the simulations, I'm not sure how would moving it to the PolicyActor will help.

Reply:
It would move the accounting in the stats. Currently every policy is modified to record the stats with your metrics. Instead you could record the penalty time outside of those policies and not affect them. If you need it in a policy, it can use it as it’s available in the event as desired.

Reply:
okay, I'll try and change it now.

Reply:
Okay, I changed the penalties to work via the PolicyActor as you suggested, but still, I do think that the logic of the filtering should be the way I did it since it keeps only the policies that will work properly with the type of event that the trace reader provides and allows comparing results against policies that do not factor in all of the provided meta-data to make a decision.

If you think this filtering is not appropriate for all use cases, I think that the filtering logic should be controlled in some way, my current idea is still to add some attribute to the configuration file to control the filtering logic.

currently, I don't really see a reason we would want to filter out policies that will work properly when using some trace reader.